"use strict";(self.webpackChunk=self.webpackChunk||[]).push([[65860],{603905:(e,o,t)=>{t.d(o,{Zo:()=>p,kt:()=>u});var r=t(667294);function n(e,o,t){return o in e?Object.defineProperty(e,o,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[o]=t,e}function a(e,o){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);o&&(r=r.filter((function(o){return Object.getOwnPropertyDescriptor(e,o).enumerable}))),t.push.apply(t,r)}return t}function i(e){for(var o=1;o<arguments.length;o++){var t=null!=arguments[o]?arguments[o]:{};o%2?a(Object(t),!0).forEach((function(o){n(e,o,t[o])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):a(Object(t)).forEach((function(o){Object.defineProperty(e,o,Object.getOwnPropertyDescriptor(t,o))}))}return e}function d(e,o){if(null==e)return{};var t,r,n=function(e,o){if(null==e)return{};var t,r,n={},a=Object.keys(e);for(r=0;r<a.length;r++)t=a[r],o.indexOf(t)>=0||(n[t]=e[t]);return n}(e,o);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);for(r=0;r<a.length;r++)t=a[r],o.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(n[t]=e[t])}return n}var l=r.createContext({}),s=function(e){var o=r.useContext(l),t=o;return e&&(t="function"==typeof e?e(o):i(i({},o),e)),t},p=function(e){var o=s(e.components);return r.createElement(l.Provider,{value:o},e.children)},c={inlineCode:"code",wrapper:function(e){var o=e.children;return r.createElement(r.Fragment,{},o)}},m=r.forwardRef((function(e,o){var t=e.components,n=e.mdxType,a=e.originalType,l=e.parentName,p=d(e,["components","mdxType","originalType","parentName"]),m=s(t),u=n,h=m["".concat(l,".").concat(u)]||m[u]||c[u]||a;return t?r.createElement(h,i(i({ref:o},p),{},{components:t})):r.createElement(h,i({ref:o},p))}));function u(e,o){var t=arguments,n=o&&o.mdxType;if("string"==typeof e||n){var a=t.length,i=new Array(a);i[0]=m;var d={};for(var l in o)hasOwnProperty.call(o,l)&&(d[l]=o[l]);d.originalType=e,d.mdxType="string"==typeof e?e:n,i[1]=d;for(var s=2;s<a;s++)i[s]=t[s];return r.createElement.apply(null,i)}return r.createElement.apply(null,t)}m.displayName="MDXCreateElement"},109842:(e,o,t)=>{t.r(o),t.d(o,{assets:()=>s,contentTitle:()=>d,default:()=>m,frontMatter:()=>i,metadata:()=>l,toc:()=>p});t(667294);var r=t(603905);function n(){return n=Object.assign||function(e){for(var o=1;o<arguments.length;o++){var t=arguments[o];for(var r in t)Object.prototype.hasOwnProperty.call(t,r)&&(e[r]=t[r])}return e},n.apply(this,arguments)}function a(e,o){if(null==e)return{};var t,r,n=function(e,o){if(null==e)return{};var t,r,n={},a=Object.keys(e);for(r=0;r<a.length;r++)t=a[r],o.indexOf(t)>=0||(n[t]=e[t]);return n}(e,o);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);for(r=0;r<a.length;r++)t=a[r],o.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(n[t]=e[t])}return n}const i={id:"error-handling",title:"Gest\xe3o de erros"},d=void 0,l={unversionedId:"Concepts/error-handling",id:"version-20-R5/Concepts/error-handling",title:"Gest\xe3o de erros",description:"O manejo de erros \xe9 o processo de antecipar e responder aos erros que possam ocorrer em sua aplica\xe7\xe3o. 4D oferece assist\xeancia completa \xe0 detec\xe7\xe3o e notifica\xe7\xe3o de erros no tempo de execu\xe7\xe3o, assim como a an\xe1lise de suas condi\xe7\xf5es.",source:"@site/i18n/pt/docusaurus-plugin-content-docs/version-20-R5/Concepts/error-handling.md",sourceDirName:"Concepts",slug:"/Concepts/error-handling",permalink:"/docs/pt/Concepts/error-handling",draft:!1,editUrl:"https://github.com/4d/docs/issues/new?title=Comment%20on%20Concepts%2Ferror-handling.md%20(20-R5)&body=Please%20enter%20your%20comment%3A",tags:[],version:"20-R5",frontMatter:{id:"error-handling",title:"Gest\xe3o de erros"},sidebar:"docs",previous:{title:"Fluxo de controlo",permalink:"/docs/pt/Concepts/control-flow"},next:{title:"Modos interpretado e compilado",permalink:"/docs/pt/Concepts/interpreted-compiled"}},s={},p=[{value:"Erro ou status",id:"erro-ou-status",level:2},{value:"Instala\xe7\xe3o de um m\xe9todo de gest\xe3o de erros",id:"instala\xe7\xe3o-de-um-m\xe9todo-de-gest\xe3o-de-erros",level:2},{value:"Alcance e componentes",id:"alcance-e-componentes",level:3},{value:"Manejo de erros dentro do m\xe9todo",id:"manejo-de-erros-dentro-do-m\xe9todo",level:3},{value:"Exemplo",id:"exemplo",level:4},{value:"Utilizar um m\xe9todo de gest\xe3o de erro vazio",id:"utilizar-um-m\xe9todo-de-gest\xe3o-de-erro-vazio",level:3},{value:"Try(expression)",id:"tryexpression",level:2},{value:"Exemplos",id:"exemplos",level:3},{value:"Try...Catch...End try",id:"trycatchend-try",level:2},{value:"Exemplo",id:"exemplo-1",level:3}],c={toc:p};function m(e){var{components:o}=e,i=a(e,["components"]);return(0,r.kt)("wrapper",n({},c,i,{components:o,mdxType:"MDXLayout"}),(0,r.kt)("p",null,"O manejo de erros \xe9 o processo de antecipar e responder aos erros que possam ocorrer em sua aplica\xe7\xe3o. 4D oferece assist\xeancia completa \xe0 detec\xe7\xe3o e notifica\xe7\xe3o de erros no tempo de execu\xe7\xe3o, assim como a an\xe1lise de suas condi\xe7\xf5es."),(0,r.kt)("p",null,"Manejo de erros responde \xe0 duas necessidades principais:"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"descobrir e consertar erros potenciais e bugs no c\xf3digo durante a fase de desenvolvimento,"),(0,r.kt)("li",{parentName:"ul"},"detectar e recuperar de erros inesperados nas aplica\xe7\xf5es implementadas; em particular pode substituir di\xe1logos de erros de sistemas (disco cheio, arquivo faltando, etc) com sua pr\xf3pria interface.")),(0,r.kt)("p",null,"Basicamente, h\xe1 duas maneiras de lidar com erros em 4D. Pode:"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("a",n({parentName:"li"},{href:"#installing-an-error-handling-method"}),"install an error-handling method"),", or"),(0,r.kt)("li",{parentName:"ul"},"use a ",(0,r.kt)("a",n({parentName:"li"},{href:"#tryexpression"}),(0,r.kt)("inlineCode",{parentName:"a"},"Try()")," keyword")," or a ",(0,r.kt)("a",n({parentName:"li"},{href:"#trycatchend-try"}),(0,r.kt)("inlineCode",{parentName:"a"},"Try/Catch")," structure")," before pieces of code that call a function, method, or expression that can throw an error.")),(0,r.kt)("p",null,":::dica Boa pr\xe1tica"),(0,r.kt)("blockquote",null,(0,r.kt)("blockquote",{parentName:"blockquote"},(0,r.kt)("p",{parentName:"blockquote"},"\xc9 recomendado instalar um m\xe9todo de gerenciamento de erros em 4D Server, para todos os c\xf3digos rodando no servidor. Quando 4D Server n\xe3o est\xe1 sendo executado ",(0,r.kt)("a",n({parentName:"p"},{href:"/docs/pt/Admin/cli"}),"headless")," (ou seja, \xe9 iniciado com sua ",(0,r.kt)("a",n({parentName:"p"},{href:"/docs/pt/ServerWindow/overview"}),"janela de administra\xe7\xe3o"),"), esse m\xe9todo evitaria que caixas de di\xe1logo inesperadas fossem exibidas na m\xe1quina do servidor. No modo headless, os erros s\xe3o registados no ficheiro ",(0,r.kt)("a",n({parentName:"p"},{href:"/docs/pt/Debugging/debugLogFiles#4ddebuglogtxt-standard"}),"4DDebugLog")," para an\xe1lise posterior."))),(0,r.kt)("p",null,":::"),(0,r.kt)("h2",n({},{id:"erro-ou-status"}),"Erro ou status"),(0,r.kt)("p",null,"Muitas fun\xe7\xf5es de classe 4D, tais como ",(0,r.kt)("inlineCode",{parentName:"p"},"entity.save()")," ou ",(0,r.kt)("inlineCode",{parentName:"p"},"transporter.send()"),', devolvem um objecto com o estatuto **. Este objecto \xe9 utilizado para armazenar erros "previs\xedveis" no contexto do tempo de execu\xe7\xe3o, por exemplo, palavra-passe inv\xe1lida, entidade bloqueada, etc., que n\xe3o interrompem a execu\xe7\xe3o do programa. Esta categoria de erros pode ser tratada por c\xf3digo normal.'),(0,r.kt)("p",null,'Outros erros "imprevis\xedveis" incluem erro de grava\xe7\xe3o em disco, falha de rede, ou em geral qualquer interrup\xe7\xe3o inesperada. Esta categoria de erros gera exce\xe7\xf5es e precisa ser tratada atrav\xe9s de um m\xe9todo de manipula\xe7\xe3o de erros ou uma palavra-chave ',(0,r.kt)("inlineCode",{parentName:"p"},"Try()"),"."),(0,r.kt)("h2",n({},{id:"instala\xe7\xe3o-de-um-m\xe9todo-de-gest\xe3o-de-erros"}),"Instala\xe7\xe3o de um m\xe9todo de gest\xe3o de erros"),(0,r.kt)("p",null,"Em 4D, todos os erros podem ser apanhados e tratados num m\xe9todo de projecto espec\xedfico, o m\xe9todo ",(0,r.kt)("strong",{parentName:"p"},"de tratamento de erros")," (ou ",(0,r.kt)("strong",{parentName:"p"},"de captura de erros"),")."),(0,r.kt)("p",null,"Uma vez instalados, os manipuladores de erros s\xe3o automaticamente chamados em modo interpretado ou compilado em caso de erro na aplica\xe7\xe3o 4D ou num dos seus componentes. Um manipulador de erros diferente pode ser chamado em fun\xe7\xe3o do contexto de execu\xe7\xe3o (ver abaixo)."),(0,r.kt)("p",null,"Para ",(0,r.kt)("em",{parentName:"p"},"instalar")," um m\xe9todo de projecto de tratamento de erros, basta chamar o comando ",(0,r.kt)("a",n({parentName:"p"},{href:"https://doc.4d.com/4dv19/help/command/en/page155.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"ON ERR CALL"))," com o nome do m\xe9todo de projecto e (opcionalmente) o \xe2mbito como par\xe2metros. Por exemplo:"),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("IO_ERRORS") //Instala o m\xe9todo de gest\xe3o de erros\n')),(0,r.kt)("p",null,"Para deixar de detectar erros e devolver o controle a 4D, chame a ",(0,r.kt)("inlineCode",{parentName:"p"},"ON ERR CALL")," com uma string vazia:"),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("";ek local) // d\xe1 o controle para o processo local\n')),(0,r.kt)("p",null,"The  ",(0,r.kt)("a",n({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page704.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"Method called on error"))," command allows you to know the name of the method installed by ",(0,r.kt)("inlineCode",{parentName:"p"},"ON ERR CALL")," for the current process. \xc9 particularmente \xfatil no contexto dos componentes porque permite mudar temporariamente e depois restaurar o m\xe9todo de captura de erros do banco de dados local:"),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),' $methCurrent:=Method called on error\n ON ERR CALL("NewMethod")\n  //Se o documento n\xe3o pouder ser aberto, um erro \xe9 gerado\n $ref:=Open document("MyDocument")\n  //Reinstala\xe7\xe3o do m\xe9todo anterior\n ON ERR CALL($methCurrent)\n\n')),(0,r.kt)("h3",n({},{id:"alcance-e-componentes"}),"Alcance e componentes"),(0,r.kt)("p",null,"Um m\xe9todo de tratamento de erros pode ser definido para diferentes contextos de execu\xe7\xe3o:"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},'ON ERR CALL("") //devolve o controle a 4D'),(0,r.kt)("li",{parentName:"ul"},"o comando ",(0,r.kt)("inlineCode",{parentName:"li"},"GET LAST ERROR STACK")," que devolve informa\xe7\xe3o sobre a pilha de erros atual da aplica\xe7\xe3o 4D."),(0,r.kt)("li",{parentName:"ul"},"In the case of an error generated by a ",(0,r.kt)("strong",{parentName:"li"},"component"),", the ",(0,r.kt)("inlineCode",{parentName:"li"},"ON ERR CALL")," error-handling method of the host application is not called, and vice versa.")),(0,r.kt)("p",null,"Exemplos:"),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("IO_Errors";ek local) //Instala um m\xe9todo local de tratamento de erros ON ERR CALL("globalHandler";ek global) //Instala um m\xe9todo global de tratamento de erros ON ERR CALL("componentHandler";ek erros de componentes) //Instala um m\xe9todo de tratamento de erros de componentes\n')),(0,r.kt)("p",null,'Pode instalar um gerenciador de erros global que servir\xe1 como "fallback" e m\xe9todos de erros locais espec\xedficos para certos processos. Um manipulador de erros global \xe9 tamb\xe9m \xfatil no servidor para evitar di\xe1logos de erro no servidor quando executado com interface.'),(0,r.kt)("p",null,"Pode definir um \xfanico m\xe9todo de captura de erros para toda a aplica\xe7\xe3o ou diferentes m\xe9todos por m\xf3dulo de aplica\xe7\xe3o. Entretanto, apenas um m\xe9todo pode ser instalado por processo."),(0,r.kt)("p",null,"Quando ocorre um erro, apenas um m\xe9todo \xe9 chamado, como descrito no diagrama seguinte:"),(0,r.kt)("p",null,(0,r.kt)("img",{alt:"o comando &lt;code&gt;GET LAST ERROR STACK&lt;/code&gt; que devolve informa\xe7\xe3o sobre a pilha de erros atual da aplica\xe7\xe3o 4D.",src:t(250161).Z,width:"1195",height:"578"})),(0,r.kt)("h3",n({},{id:"manejo-de-erros-dentro-do-m\xe9todo"}),"Manejo de erros dentro do m\xe9todo"),(0,r.kt)("p",null,"Within the custom error method, you have access to several pieces of information that will help you identifying the error:"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("p",{parentName:"li"},"vari\xe1veis sistema (*):"),(0,r.kt)("ul",{parentName:"li"},(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("inlineCode",{parentName:"li"},"Error")," (inteiro longo): c\xf3digo de erro"),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("inlineCode",{parentName:"li"},"Error method"),"(texto): nome do m\xe9todo que provocou o erro"),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("inlineCode",{parentName:"li"},"Error line")," (entero largo): n\xfamero de l\xednea do m\xe9todo que provocou o erro"),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("inlineCode",{parentName:"li"},"Error formula")," (text): f\xf3rmula do c\xf3digo 4D (texto bruto) que est\xe1 na origem do erro.")))),(0,r.kt)("admonition",n({},{type:"info"}),(0,r.kt)("p",{parentName:"admonition"},"(",(0,r.kt)("em",{parentName:"p"},") 4D mant\xe9m automaticamente uma s\xe9rie de vari\xe1veis chamadas ",(0,r.kt)("strong",{parentName:"em"},"vari\xe1veis sistema"),", que respondem a diferentes necessidades. Consulte o "),"Manual da linguagem de 4D*.")),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"o comando ",(0,r.kt)("a",n({parentName:"li"},{href:"https://doc.4d.com/4dv19/help/command/en/page1799.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"\xfaltimos erros"))," que devolve uma cole\xe7\xe3o da pilha actual de erros que ocorreram na aplica\xe7\xe3o 4D. Tamb\xe9m pode usar o comando ",(0,r.kt)("a",n({parentName:"li"},{href:"https://doc.4d.com/4dv19/help/command/en/page1015.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"GET LAST ERROR STACK"))," que devolve a mesma informa\xe7\xe3o que os arrays."),(0,r.kt)("li",{parentName:"ul"},"o comando ",(0,r.kt)("inlineCode",{parentName:"li"},"Get call chain")," que devolve uma cole\xe7\xe3o de objetos que descrevem cada passo da string de chamadas a m\xe9todos dentro do processo atual.")),(0,r.kt)("h4",n({},{id:"exemplo"}),"Exemplo"),(0,r.kt)("p",null,"Aqui est\xe1 um sistema de gest\xe3o de erros simples:"),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'//instalar o m\xe9todo de gest\xe3o de erros\n ON ERR CALL("errorMethod")\n //... executar o c\xf3digo\n ON ERR CALL("") //retorna o controle para 4D\n')),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'// m\xe9todo projeto errorMethod\n If(Error#1006) //essa n\xe3o \xe9 uma interrup\xe7\xe3o do usu\xe1rio\n    ALERT("Um erro foi produzido "+String(Error)+". O c\xf3digo em quest\xe3o \xe9: \\""+Error formula+"\\"")\n End if\n')),(0,r.kt)("h3",n({},{id:"utilizar-um-m\xe9todo-de-gest\xe3o-de-erro-vazio"}),"Utilizar um m\xe9todo de gest\xe3o de erro vazio"),(0,r.kt)("p",null,"Se quiser que a caixa de di\xe1logo fique escondida, pode instalar um m\xe9todo de gest\xe3o de erros vazio. A vari\xe1vel sistema ",(0,r.kt)("inlineCode",{parentName:"p"},"Error")," pode ser provada em qualquer m\xe9todo, ou seja, fora do m\xe9todo de gest\xe3o de erros:"),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("emptyMethod") //emptyMethod exists but is empty\n$doc:=Open document( "myFile.txt")\nIf (Error=-43)\n    ALERT("File not found.")\nEnd if\nON ERR CALL("")\n')),(0,r.kt)("h2",n({},{id:"tryexpression"}),"Try(expression)"),(0,r.kt)("p",null,"The ",(0,r.kt)("inlineCode",{parentName:"p"},"Try(expression)")," statement allows you to test a single-line expression in its actual execution context (including, in particular, local variable values) and to intercept errors it throws so that the 4D error dialog box is not displayed. Using ",(0,r.kt)("inlineCode",{parentName:"p"},"Try(expression)")," provides an easy way to handle simple error cases with a very low number of code lines, and without requiring an error-handling method."),(0,r.kt)("admonition",n({},{type:"note"}),(0,r.kt)("p",{parentName:"admonition"},"If you want to try a more complex code than a single-line expression, you might consider using a ",(0,r.kt)("a",n({parentName:"p"},{href:"#trycatchend-try"}),(0,r.kt)("inlineCode",{parentName:"a"},"Try/Catch")," structure"),".")),(0,r.kt)("p",null,"A sintaxe formal da instru\xe7\xe3o ",(0,r.kt)("inlineCode",{parentName:"p"},"Try(expression)")," \xe9:"),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),"\nTry (expression) : any | Undefined\n\n")),(0,r.kt)("p",null,(0,r.kt)("em",{parentName:"p"},"expression")," pode ser qualquer express\xe3o v\xe1lida."),(0,r.kt)("p",null,"If an error occurred during its execution, it is intercepted and no error dialog is displayed, whether an ",(0,r.kt)("a",n({parentName:"p"},{href:"#installing-an-error-handling-method"}),"error-handling method")," was installed or not before the call to ",(0,r.kt)("inlineCode",{parentName:"p"},"Try()"),". Se ",(0,r.kt)("em",{parentName:"p"},"expression")," retorna um valor, ",(0,r.kt)("inlineCode",{parentName:"p"},"Try()")," retorna o \xfaltimo valor calculado, caso contr\xe1rio, ele retorna ",(0,r.kt)("inlineCode",{parentName:"p"},"Undefined"),"."),(0,r.kt)("p",null,"Voc\xea pode tratar o(s) erro(s) usando o comando ",(0,r.kt)("a",n({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page1799.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"Last errors")),". If ",(0,r.kt)("em",{parentName:"p"},"expression")," throws an error within a stack of ",(0,r.kt)("inlineCode",{parentName:"p"},"Try()")," calls, the execution flow stops and returns to the latest executed ",(0,r.kt)("inlineCode",{parentName:"p"},"Try()")," (the first found back in the call stack)."),(0,r.kt)("admonition",n({},{type:"note"}),(0,r.kt)("p",{parentName:"admonition"},"If an ",(0,r.kt)("a",n({parentName:"p"},{href:"#installing-an-error-handling-method"}),"error-handling method")," is installed by ",(0,r.kt)("em",{parentName:"p"},"expression"),", it is called in case of error.")),(0,r.kt)("h3",n({},{id:"exemplos"}),"Exemplos"),(0,r.kt)("ol",null,(0,r.kt)("li",{parentName:"ol"},"You want to display the contents of a file if the file can be open without error, and if its contents can be read. Voc\xea pode escrever:")),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'var $text : Text\nvar $file : 4D.File := File("/RESOURCES/myFile.txt")\nvar $fileHandle : 4D.FileHandle := Try($file.open())\nIf ($fileHandle # Null)\n  $text:=Try($fileHandle.readText()) || "Erro ao ler o arquivo"\nEnd if\n')),(0,r.kt)("ol",n({},{start:2}),(0,r.kt)("li",{parentName:"ol"},"Voc\xea deseja tratar o erro de divis\xe3o por zero. Nesse caso, voc\xea deseja retornar 0 e lan\xe7ar um erro:")),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'function divide( $p1: real; $p2: real)-> $result: real\n  if ($p2=0)\n     $result:=0 //apenas para fins de clareza (0 \xe9 o padr\xe3o para reais)\n     throw(-12345; "Division by zero")\n  else\n    $result:=$p1/$p2\n  end if\n\nfunction test()\n  $result:=Try(divide($p1;$p2))\n  If (Last errors # null)\n    ALERT("Error")\n  End if\n\n')),(0,r.kt)("ol",n({},{start:3}),(0,r.kt)("li",{parentName:"ol"},"You want to handle both ",(0,r.kt)("a",n({parentName:"li"},{href:"#error-or-status"}),"predictable and non-predictable")," errors:")),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),'var $e:=ds.Employee.new()\n$e.name:="Smith"\n$status:=Try($e.save()) // Captura de erros previs\xedveis e n\xe3o previs\xedveis\nIf ($status.success)\n   ALERT( "Success")\nElse\n   ALERT( "Error: "+JSON Stringify($status.errors))\nEnd if\n\n')),(0,r.kt)("h2",n({},{id:"trycatchend-try"}),"Try...Catch...End try"),(0,r.kt)("p",null,"The ",(0,r.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," structure allows you to test a block code in its actual execution context (including, in particular, local variable values) and to intercept errors it throws so that the 4D error dialog box is not displayed."),(0,r.kt)("p",null,"Unlike the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try(expression)")," keyword that evaluates a single-line expression, the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," structure allows you to evaluate any code block, from the most simple to the most complex, without requiring an error-handling method. In addition, the ",(0,r.kt)("inlineCode",{parentName:"p"},"Catch")," block can be used to handle the error in any custom way."),(0,r.kt)("p",null,"The formal syntax of the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," structure is:"),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),"\nTry \n    statement(s) // Code to evaluate\nCatch\n    statement(s) // Code to execute in case of error\nEnd try\n\n")),(0,r.kt)("p",null,"The code placed between the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try")," and the ",(0,r.kt)("inlineCode",{parentName:"p"},"Catch")," keywords is first executed, then the flow depends on the error(s) encountered during this execution."),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"If no error is thrown, the code execution continues after the corresponding ",(0,r.kt)("inlineCode",{parentName:"li"},"End try")," keyword. The code placed between the ",(0,r.kt)("inlineCode",{parentName:"li"},"Catch")," and the ",(0,r.kt)("inlineCode",{parentName:"li"},"End try")," keywords is not executed."),(0,r.kt)("li",{parentName:"ul"},"If the code block execution throws a ",(0,r.kt)("em",{parentName:"li"},"non-deferred error"),", the execution flow stops and executes the corresponding ",(0,r.kt)("inlineCode",{parentName:"li"},"Catch")," code block."),(0,r.kt)("li",{parentName:"ul"},"If the code block execution throws a ",(0,r.kt)("em",{parentName:"li"},"deferred error"),", the execution flow continues until the end of the ",(0,r.kt)("inlineCode",{parentName:"li"},"Try")," block and then executes the corresponding ",(0,r.kt)("inlineCode",{parentName:"li"},"Catch")," code block.")),(0,r.kt)("admonition",n({},{type:"note"}),(0,r.kt)("p",{parentName:"admonition"},"If a ",(0,r.kt)("em",{parentName:"p"},"deferred")," error is thrown outside of the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try")," block, the code execution continues until the end of the method or function.")),(0,r.kt)("admonition",n({},{type:"info"}),(0,r.kt)("p",{parentName:"admonition"},"For more information on ",(0,r.kt)("em",{parentName:"p"},"deferred")," and ",(0,r.kt)("em",{parentName:"p"},"non-deferred")," errors, please refer to the ",(0,r.kt)("a",n({parentName:"p"},{href:"https://doc.4d.com/4dv20R/help/command/en/page1805.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"throw"))," command description.")),(0,r.kt)("p",null,"In the ",(0,r.kt)("inlineCode",{parentName:"p"},"Catch")," code block, you can handle the error(s) using standard error handling commands. The ",(0,r.kt)("a",n({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page1799.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"Last errors"))," function contains the last errors collection. You can ",(0,r.kt)("a",n({parentName:"p"},{href:"#installing-an-error-handling-method"}),"declare an error-handling method")," in this code block, in which case it is called in case of error (otherwise the 4D error dialog box is displayed)."),(0,r.kt)("admonition",n({},{type:"note"}),(0,r.kt)("p",{parentName:"admonition"},"If an ",(0,r.kt)("a",n({parentName:"p"},{href:"#installing-an-error-handling-method"}),"error-handling method")," is installed in the code placed between the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try")," and the ",(0,r.kt)("inlineCode",{parentName:"p"},"Catch")," keywords, it is called in case of error.")),(0,r.kt)("h3",n({},{id:"exemplo-1"}),"Exemplo"),(0,r.kt)("p",null,"Combining transactions and ",(0,r.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," structures allows writing secured code for critical features."),(0,r.kt)("pre",null,(0,r.kt)("code",n({parentName:"pre"},{className:"language-4d"}),"Function createInvoice($customer : cs.customerEntity; $items : Collection; $invoiceRef : Text) : cs.invoiceEntity\n    var $newInvoice : cs.invoiceEntity\n    var $newInvoiceLine : cs.invoiceLineEntity\n    var $item : Object\n    ds.startTransaction()\n    Try\n        $newInvoice:=This.new()\n        $newInvoice.customer:=$customer\n        $newInvoice.invoiceRef:=$invoiceRef\n        For each ($item; $items)\n            $newInvoiceLine:=ds.invoiceLine.new()\n            $newInvoiceLine.item:=$item.item\n            $newInvoiceLine.amount:=$item.amount\n            $newInvoiceLine.invoice:=$newInvoice\n            //call other specific functions to validate invoiceline\n            $newInvoiceLine.save()\n        End for each \n        $newInvoice.save()\n        ds.validateTransaction()\n    Catch\n        ds.cancelTransaction()\n        ds.logErrors(Last errors)\n        $newInvoice:=Null\n    End try\n    return $newInvoice\n\n")))}m.isMDXComponent=!0},250161:(e,o,t)=>{t.d(o,{Z:()=>r});const r=t.p+"assets/images/error-schema-973d40c4618e79de87f884622e4e930f.png"}}]);